flash_io.c

```diff
@@ -0,0 +1,197 @@
+#include <hardware/flash.h>
+#include <hardware/sync.h>
+
+#include <stdio.h>
+#include <string.h>
+
+/*
+ *   The main purpose of saving user data to flash memory is to
+ *   handle unexpected loss of electrical power.
+ *   In such a case, the furnace driver can boot itself up
+ *   with previous config.
+ *
+ *   In order to read from flash,
+ *   config (both size and initialized targets) must be matching.
+ *
+ *   Also, we are reading from flash if and only if
+ *   the last record is the same as current config.
+ *   It is done in order to NOT boot with obsolete data.
+ *
+ *  memory layout:
+ *    size
+ *    targets
+ *    pilot.des_temp
+ *    pilot.is_enabled
+ *    log_bits
+ *    pwm_water
+ *
+ *    It is worth to note that you cannot rewrite flash memory,
+ *    once you write something in it,
+ *    you cannot overwrite this without erasing whole flash sector.
+ *    It is not described in pico-sdk.
+ *    It results from NOR/NAND flash memory.
+ */
+
+enum __attribute__((packed))
+init_targets {
+  MAGNETRON = 1 << 0,
+  WATER     = 1 << 1
+};
+
+// get symbols from the linker
+extern size_t
+FLASH_USER_DATA_START;
+
+extern size_t
+FLASH_USER_DATA_END;
+
+static const uint8_t
+FLASH_USER_DATA_LEN = sizeof(flash_io_t);
+
+static const void*
+flash_ptr;
+
+static const enum init_targets
+targets = 0
+  #if CONFIG_MAGNETRON
+    + MAGNETRON
+  #endif
+  #if CONFIG_WATER
+    + WATER
+  #endif
+;
+
+static bool
+check_end_of_user_flash()
+{
+  if(flash_ptr == &FLASH_USER_DATA_END)
+  {
+    const uint32_t ints = save_and_disable_interrupts();
+    flash_range_erase((uint32_t)&FLASH_USER_DATA_START - XIP_BASE, FLASH_SECTOR_SIZE);
```

Comment:
What's `XIP_BASE`? Don't we underflow when we subtract from `FLASH_USER_DATA_START`?

Reply:
XIP_BASE is defined as 0x10000000 (from rp2040 docs).
The same address is origin for FLASH memory in mem_map.ld.
[This article](https://kevinboone.me/picoflash.html) says the same.
So when we subtract XIP_BASE from FLASH_USER_DATA_START, we will get offset in the flash memory,
as needed in the rp2040 [API for flash_range_program](https://www.raspberrypi.com/documentation/pico-sdk/hardware.html#rpip8ee511575881aa0f3936)